File: src/common/timing.f90
module timing_module
    implicit none
    
    integer(kind=8) :: count_rate
    integer(kind=8) :: start_math, end_math
    integer(kind=8) :: start_io, end_io
    real(kind=8)    :: total_math_time = 0.0d0
    real(kind=8)    :: total_io_time = 0.0d0
    integer(kind=8) :: total_steps = 0

contains

    subroutine init_timer()
        call system_clock(count_rate=count_rate)
        total_math_time = 0.0d0
        total_io_time = 0.0d0
        total_steps = 0
        print *, "Hardware Timer Initialized. Clock Rate: ", count_rate
    end subroutine init_timer

    subroutine start_math_timer()
        call system_clock(count=start_math)
    end subroutine start_math_timer

    subroutine stop_math_timer()
        call system_clock(count=end_math)
        total_math_time = total_math_time + real(end_math - start_math, 8) / real(count_rate, 8)
        total_steps = total_steps + 1
    end subroutine stop_math_timer

    subroutine start_io_timer()
        call system_clock(count=start_io)
    end subroutine start_io_timer

    subroutine stop_io_timer()
        call system_clock(count=end_io)
        total_io_time = total_io_time + real(end_io - start_io, 8) / real(count_rate, 8)
    end subroutine stop_io_timer

    subroutine print_timing_results(nx, ny, arch)
        integer, intent(in) :: nx, ny
        character(len=*), intent(in) :: arch
        real(kind=8) :: mcups
        logical :: file_exists

        if (total_math_time > 0.0d0) then
            mcups = real(nx, 8) * real(ny, 8) * real(total_steps, 8) / (total_math_time * 1.0d6)
        else
            mcups = 0.0d0
        end if

        ! --- 1. Print to Terminal ---
        print *, "------------------------------------------------"
        print *, "               BENCHMARK RESULTS                "
        print *, "------------------------------------------------"
        print *, "Architecture:      ", trim(arch)
        print *, "Resolution:        ", nx, "x", ny
        print *, "Total Steps:       ", total_steps
        print *, "Math Time (Compute): ", total_math_time, " seconds"
        print *, "I/O Time (Disk/PCIe):", total_io_time, " seconds"
        print *, "Total Wall-Clock:  ", total_math_time + total_io_time, " seconds"
        print *, "Performance:       ", mcups, " MCUPS"
        print *, "------------------------------------------------"

        ! --- 2. Append to CSV ---
        inquire(file="benchmarks.csv", exist=file_exists)
        if (file_exists) then
            open(unit=20, file="benchmarks.csv", position="append", status="old", action="write")
        else
            open(unit=20, file="benchmarks.csv", status="new", action="write")
            write(20, '(A)') "Language,Architecture,GridSize,MathTime_sec,IOTime_sec,MCUPS"
        end if        
        write(20, '(A,A,A,A,I0,A,F0.6,A,F0.6,A,F0.6)') &
            "Fortran", ",", trim(arch), ",", nx, ",", total_math_time, ",", total_io_time, ",", mcups
        close(20)

    end subroutine print_timing_results

end module timing_module

File: src/common/timing.cpp
#include "timing.hpp"
#include <iostream>
#include <iomanip>
#include <chrono>
#include <fstream>  
#include <string>  

namespace TimingModule {
    using namespace std::chrono;
    
    // Internal timing variables (hidden from main.cpp)
    static high_resolution_clock::time_point start_math;
    static high_resolution_clock::time_point end_math;
    static high_resolution_clock::time_point start_io;
    static high_resolution_clock::time_point end_io;

    static double total_math_time = 0.0;
    static double total_io_time = 0.0;
    static long long total_steps = 0;

    void init_timer() {
        total_math_time = 0.0;
        total_io_time = 0.0;
        total_steps = 0;
        std::cout << "Hardware Timer Initialized (std::chrono)." << std::endl;
    }

    void start_math_timer() {
        start_math = high_resolution_clock::now();
    }

    void stop_math_timer() {
        end_math = high_resolution_clock::now();
        duration<double> time_span = duration_cast<duration<double>>(end_math - start_math);
        total_math_time += time_span.count();
        total_steps++;
    }

    void start_io_timer() {
        start_io = high_resolution_clock::now();
    }

    void stop_io_timer() {
        end_io = high_resolution_clock::now();
        duration<double> time_span = duration_cast<duration<double>>(end_io - start_io);
        total_io_time += time_span.count();
    }

    void print_timing_results(int nx, int ny, const std::string& arch) {
        double mcups = 0.0;
        if (total_math_time > 0.0) {
            // Calculate MCUPS: (nx * ny * total_steps) / (math_time * 10^6)
            mcups = (static_cast<double>(nx) * static_cast<double>(ny) * static_cast<double>(total_steps)) / (total_math_time * 1.0e6);
        }

        // --- 1. Print to Standard Terminal ---
        std::cout << "------------------------------------------------\n";
        std::cout << "               BENCHMARK RESULTS                \n";
        std::cout << "------------------------------------------------\n";
        std::cout << "Architecture:      " << arch << "\n";
        std::cout << "Resolution:        " << nx << "x" << ny << "\n";
        std::cout << "Total Steps:       " << total_steps << "\n";
        std::cout << std::fixed << std::setprecision(6);
        std::cout << "Math Time (Compute): " << total_math_time << " seconds\n";
        std::cout << "I/O Time (Disk/IO):  " << total_io_time << " seconds\n";
        std::cout << "Total Wall-Clock:  " << (total_math_time + total_io_time) << " seconds\n";
        std::cout << std::setprecision(4);
        std::cout << "Performance:       " << mcups << " MCUPS\n";
        std::cout << "------------------------------------------------\n";

        // --- 2. Append to CSV for Automation ---
        std::string filename = "benchmarks.csv";
        
        // Check if the file already exists
        std::ifstream file_check(filename);
        bool file_exists = file_check.good();
        file_check.close();

        // Open file in append mode
        std::ofstream csv_file(filename, std::ios::app);
        if (csv_file.is_open()) {
            // If the file is brand new, write the CSV headers first
            if (!file_exists) {
                csv_file << "Language,Architecture,GridSize,MathTime_sec,IOTime_sec,MCUPS\n";
            }

            // Write the data row with strict formatting
            csv_file << std::fixed << std::setprecision(6);
            csv_file << "C++," << arch << "," << nx << "," 
                     << total_math_time << "," << total_io_time << "," 
                     << mcups << "\n";
                     
            csv_file.close();
        } else {
            std::cerr << "Warning: Could not open " << filename << " to write benchmark data.\n";
        }
    }
}